src/libraries/System.IO.Compression/src/System/IO/Compression/ZipCryptoStream.cs

```diff
@@ -0,0 +1,254 @@
+            _position += 12;
+        }
+
+        private void InitKeysFromBytes(ReadOnlySpan<char> password)
+        {
+            _key0 = 305419896;
+            _key1 = 591751049;
+            _key2 = 878082192;
+
+            // ZipCrypto uses raw bytes; ASCII is the most interoperable (UTF8 also acceptable).
+            var bytes = password.ToArray();
+            foreach (byte b in bytes)
+                UpdateKeys(b);
+        }
+
+        private void ValidateHeader(byte expectedCheckByte)
+        {
+            byte[] hdr = new byte[12];
+            int read = 0;
+            while (read < hdr.Length)
+            {
+                int n = _base.Read(hdr, read, hdr.Length - read);
+                if (n <= 0) throw new InvalidDataException("Truncated ZipCrypto header.");
+                read += n;
+            }
+
+            for (int i = 0; i < hdr.Length; i++)
+                hdr[i] = DecryptByte(hdr[i]);
+
+            if (hdr[11] != expectedCheckByte)
+                throw new InvalidDataException("Invalid password for encrypted ZIP entry.");
+        }
+
+        private void UpdateKeys(byte b)
+        {
+            _key0 = Crc32Update(_key0, b);
+            _key1 += (_key0 & 0xFF);
+            _key1 = _key1 * 134775813 + 1;
+            _key2 = Crc32Update(_key2, (byte)(_key1 >> 24));
+        }
+
+        private byte DecipherByte()
+        {
+            uint temp = _key2 | 2; // use uint to avoid narrowing issues
+            return (byte)((temp * (temp ^ 1)) >> 8);
+        }
+
+        private byte DecryptByte(byte ciph)
+        {
+            byte m = DecipherByte();
+            byte plain = (byte)(ciph ^ m);
+            UpdateKeys(plain);
+            return plain;
+        }
+
+        // ---- Stream overrides ----
+
+        public override bool CanRead => !_encrypting;
+        public override bool CanSeek => false;
+        public override bool CanWrite => _encrypting;
+        public override long Length => throw new NotSupportedException();
+        public override long Position
+        {
+            get => _position;
+            set => throw new NotSupportedException("ZipCryptoStream does not support seeking.");
+        }
+        public override void Flush() => _base.Flush();
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (!_encrypting)
+            {
+                ArgumentNullException.ThrowIfNull(buffer);
+                if ((uint)offset > buffer.Length || (uint)count > buffer.Length - offset)
+                    throw new ArgumentOutOfRangeException();
+
+                int n = _base.Read(buffer, offset, count);
+                for (int i = 0; i < n; i++)
+                    buffer[offset + i] = DecryptByte(buffer[offset + i]);
+                _position += n;
+                return n;
+            }
+            throw new NotSupportedException("Stream is in encryption (write-only) mode.");
+        }
```

Comment:
You can minimize the amount of code by redirecting to a more flexible overloads

```suggestion
        public override int Read(byte[] buffer, int offset, int count)
        {
                ValidateBufferArguments(buffer, offset, count);
                return Read(buffer.AsSpan(offset, count));
        }
```

You can do similar trick for the Write method